### PR TITLE
Notebook sur des exemples et tests de svd et dpp

### PR DESCRIPTION
**Pull request by fravoi**
Notebook permettant de donner des exemples pratiques après svd et dpp (graphiques, exemples, tests)

### CODE REVIEW

File: mangaki/mangaki/utils/dpp.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
+from sklearn.utils.extmath import randomized_svd
+from scipy.spatial.distance import pdist, squareform
+from mangaki.utils.svd import MangakiSVD
+from mangaki.utils.values import rating_values
+from mangaki.models import Rating
+import pandas, random
+import numpy as np
+
+class SimilarityMatrix(object):
+
```
Comment (Elarnon) [inline mangaki/mangaki/utils/dpp.py]: Cette classe devrait avoir un constructeur (`__init__`) qui effectue la création de la matrice + matrice de similarité en prenant les options nécessaires.

Je suggère la signature suivante : `def __init__(self, fname=None, metric='cosine'):` avec la convention que si `fname` est `None`, alors les valeurs seront chargées depuis la base de donnée.

File: mangaki/mangaki/utils/dpp.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+from mangaki.utils.svd import MangakiSVD
+from mangaki.utils.values import rating_values
+from mangaki.models import Rating
+import pandas, random
+import numpy as np
+
+class SimilarityMatrix(object):
+	
+#option à rajouter : nb_ratins, nb_user ou nb_items ? 
+	#option : database' or 'csv'
```
Comment (Elarnon) [inline mangaki/mangaki/utils/dpp.py]: Enlever ces commentaires.

File: mangaki/mangaki/utils/dpp.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	def make_matrix(self, option_data):
```
Comment (Elarnon) [inline mangaki/mangaki/utils/dpp.py]: `build_matrix` serait un nom plus approprié, cf remarque sur `__init__` plus haut. `option_data` est hautement non explicite comme nom d'argument, je propose à nouveau `fname=None` en chargeant soit depuis le fichier situé à `fname`, soit depuis la base de données si `fname` est `None`, e.g.:

```python
if fname is None:
    content = Rating.objects.values_list('user_id', 'work_id', 'choice')
    # ...
else:
    content = pandas.read_csv(fname, header=None).as_matrix()
    #...
```

File: mangaki/mangaki/utils/dpp.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+		user_set, item_set=set(), set()
+		if option_data=='database':
+			content=Rating.objects.values_list('user_id', 'work_id', 'choice')
+			
+			#user_items=[(content[i][0], content[i][1]) for i in range(len(content))]
```
Comment (Elarnon) [inline mangaki/mangaki/utils/dpp.py]: Merci de ne pas laisser de code commenté.

File: mangaki/mangaki/utils/dpp.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+			content = pandas.read_csv('../data/ratings.csv', header=None).as_matrix()
+			for user_id, item_id in content[:,0:2]:
+				user_set.add(user_id)
+				item_set.add(item_id) 
+		else:
+			return "Erreur : les options sont 'database' ou 'csv'"           
+        
+		user_dict= {v:k for k,v in enumerate(user_set)}
+		item_dict= {v:k for k,v in enumerate(item_set)}
+		matrix = np.zeros((len(user_set), len(item_set)), dtype=np.float64)
```
Comment (Elarnon) [inline mangaki/mangaki/utils/dpp.py]: Utiliser plutôt une matrice creuse ici.

File: mangaki/mangaki/utils/dpp.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	
+class MangakiUniform(object):
+	
+	def __init__(self, nb_points):
+		self.nb_points = nb_points
+
+
+	def sample_k(self,items, matrix_similarity):
+		uniform_items = items
+		random.shuffle(uniform_items)
```
Comment (Elarnon) [inline mangaki/mangaki/utils/dpp.py]: Ceci modifie l'objet `items` passé en arguments. Utiliser plutôt [`np.random.choice(items, size=self.nb_points)`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.choice.html).

File: mangaki/mangaki/utils/dpp.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+
+#n'a pas d'intérêt au final
+#TODO : delete
+#similarity_fn='jaccard' or 'coisine'
+#def fit(self, matrix, all_dataset=False): #obtention de la matrice de similarité
+#	matrix_fit=similarity(matrix, similarity_fn)
+#	return matrix_fit
+    
+#thanks to mehdidc on github
```
Comment (Elarnon) [inline mangaki/mangaki/utils/dpp.py]: Mettre un lien vers le dépôt initial dans la docstring.

File: mangaki/mangaki/utils/dpp.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+# tirer au hasard des points
+#def sample_k(items, similiarity, k, max_nb_iterations=1000, rng=np.random):
+       #def sample_k(items, similiarity, k, max_nb_iterations=1000, rng=np.random):
+	def sample_k(self, items, L, k, max_nb_iterations=1000, rng=np.random):
+	
+	#Sample a list of k items from a DPP defined
```
Comment (Elarnon) [inline mangaki/mangaki/utils/dpp.py]: Ceci devrait être une docstring (`"""`).

File: mangaki/mangaki/utils/dpp.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+		random.shuffle(uniform_items)
+		return uniform_items[:self.nb_points]
+    	
+	def __str__(self):
+		return 'uniform list, nb_points=%d' % self.nb_points
+
+class MangakiDPP(object):
+
+	def __init__(self, nb_points):
+		self.nb_points = nb_points
```
Comment (Elarnon) [inline mangaki/mangaki/utils/dpp.py]: Pourquoi `MangakiDPP` définit-il `self.nb_points` pour ne jamais l'utiliser ensuite ?

Comment (Elarnon) [inline mangaki/mangaki/utils/dpp.py]: Je pense qu'on voudrait pouvoir utiliser `MangakiDPP` de la façon suivante : 
```python
similarity = SimilarityMatrix()
dpp = MangakiDPP(similarity)
dpp.sample_k(10)
```

File: mangaki/mangaki/utils/dpp.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+#det de pdist : np.linalg.det(squareform(pdist(svd.VT[:,uniform_items].T, metric='cosine')))
+
+#diamètre d'ordre r : (ici 1 pr l'instant) 
+"""
+n = len(sampled_items)
+coefficient_sample = 2/(n*(n-1))*(pdist(svd.VT[:,sampled_items].T).sum())
+"""
+
+def compare(nb_points, nb_iterations):
```
Comment (Elarnon) [inline mangaki/mangaki/utils/dpp.py]: À nettoyer et améliorer. Il faudrait que `compare` prenne une liste d'objets (`MangakiUniform` / `MangakiDPP` / autres si on en créé plus tard) et les compare entre eux en samplant le même nombre de points etc, d'où la proposition d'avoir `sample_k` qui prend uniquement `k` en argument faite plus haut.

Proposition de signature : `def compare(similarity, algos, nb_points, nb_iterations=100):`.

Regarder [compare.py](https://github.com/mangaki/mangaki/blob/master/mangaki/mangaki/management/commands/compare.py) pour inspiration.

File: mangaki/mangaki/utils/dpp.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+                item_set.add(item_id)
+        else:
+            content = pandas.read_csv(fname,
+                                      header=None).as_matrix()
+            for user_id, item_id in content[:, 0:2]:
+                user_set.add(user_id)
+                item_set.add(item_id)
+        user_dict = {v: k for k, v in enumerate(user_set)}
+        item_dict = {v: k for k, v in enumerate(item_set)}
+        matrix = lil_matrix((len(user_set), len(item_set)), dtype=np.float64)
```
Comment (Elarnon) [inline mangaki/mangaki/utils/dpp.py]: Il vaut mieux plutôt utiliser une `csc_matrix` ici, à construire en un bloc plutôt qu'une boucle (cf ce qu'on a fait dans le notebook).

File: mangaki/mangaki/utils/dpp.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+                   item_dict[item_id]] = rating_values[rating]
+        self.user_dict = user_dict
+        self.item_dict = item_dict
+        self.user_set = user_set
+        self.item_set = item_set
+        return matrix
+
+    def make_svd_matrix(self, matrix):
+        self.U, self.sigma, self.VT = randomized_svd(matrix,
+                                                     10,
```
Comment (Elarnon) [inline mangaki/mangaki/utils/dpp.py]: Il faudrait pouvoir choisir combien de composantes utiliser dans le `__init__`. Par ailleurs, utiliser les arguments mots-clefs rendrait plus clair cet appel présentemment cryptique (`randomized_svd(matrix, n_components=10, n_oversample=10, random_state=42)`). Pourquoi définir `n_oversample`, surtout à sa valeur par défaut ?

**Comment by RaitoBezarius** [comment]
@fravoi Cette PR est encore d'actualité ?
Peut-t-on avoir une titre plus clair et une vraie description, s'il te plaît ?

Si ce n'est plus d'actualité, on peut fermer la PR.

**Comment by fravoi** [comment]
Euh en fait, il faudra que je regarde. Je peux sûrement refaire un notebook correct pour faire les graphiques après SVD et des exemples/tests sur dpp. Merci :)
